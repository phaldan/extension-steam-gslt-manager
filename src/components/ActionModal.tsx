import * as React from 'react';
import { Button, Glyphicon, Modal } from 'react-bootstrap';

interface ActionModalProps {
  buttonIcon?: string;
  buttonSize?: string;
  buttonStyle?: string;
  buttonText?: string;
  disabled?: boolean;
  title?: string;
  action?: string;
  close?: string;
  size?: string;
  onAction?: () => void;
  onOpen?: () => void;
}

interface ActionModalState {
  showModal: boolean;
}

export default class ActionModal extends React.Component<ActionModalProps, ActionModalState> {
  public static defaultProps: Partial<ActionModalProps> = {
    buttonStyle: 'success',
    close: 'Abort',
    disabled: false,
  };

  constructor(props) {
    super(props);
    this.state = { showModal: false };
  }

  public render() {
    const { showModal } = this.state;
    const {
      buttonIcon,
      buttonSize,
      buttonStyle,
      buttonText,
      disabled,
      title,
      children,
      size,
      action,
    } = this.props;

    return (
      <span>
        <Button
          bsStyle={buttonStyle}
          bsSize={buttonSize}
          onClick={this.open}
          disabled={disabled}
          className="js-open"
        >
          {buttonIcon && <Glyphicon glyph={buttonIcon} />} {buttonText}
        </Button>
        <Modal
          bsSize={size}
          show={showModal}
          onHide={this.close}
        >
          <Modal.Header closeButton={true}>
            <Modal.Title>{title}</Modal.Title>
          </Modal.Header>
          <Modal.Body>
            {children}
          </Modal.Body>
          {action && this.createFooter()}
        </Modal>
      </span>
    );
  }

  private close = () => {
    this.setState({ showModal: false });
  }

  private open = () => {
    if (!this.props.disabled) {
      if (this.props.onOpen) {
        this.props.onOpen();
      }
      this.setState({ showModal: true });
    }
  }

  private createFooter() {
    const { action, close } = this.props;
    return (
      <Modal.Footer>
        <Button onClick={this.close} className="js-close">
          {close}
        </Button>
        <Button
          bsStyle="primary"
          onClick={this.triggerActionCallback}
          className="js-action"
        >
          {action}
        </Button>
      </Modal.Footer>
    );
  }

  private triggerActionCallback = () => {
    this.close();
    if (this.state.showModal && this.props.onAction) {
       this.props.onAction();
    }
  }
}
