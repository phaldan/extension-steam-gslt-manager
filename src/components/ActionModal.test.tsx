import { shallow } from 'enzyme';
import 'jest-enzyme';
import * as React from 'react';
import ActionModal from './ActionModal';

describe('<ActionModal>', () => {
  test('style and text customizations', () => {
    const element = (
      <ActionModal
        buttonSize="small"
        buttonText="Open"
        buttonIcon="more"
        title="TitleText"
        action="TriggerCallback"
        close="Close"
      >
        <p>Lorem Ipsum</p>
      </ActionModal>
    );
    const target = shallow(element);

    const openButton = target.find('.js-open');
    expect(openButton).toHaveProp('bsSize', 'small');
    expect(openButton).toHaveProp('disabled', false);
    expect(openButton.contains('Open')).toEqual(true);
    expect(target.find('Glyphicon')).toHaveProp('glyph', 'more');

    expect(target.find('Modal')).toHaveProp('show', false);
    expect(target.find('ModalTitle').contains('TitleText')).toEqual(true);
    expect(target.find('ModalBody').contains(<p>Lorem Ipsum</p>)).toEqual(true);

    expect(target.find('.js-close').contains('Close')).toEqual(true);
    expect(target.find('.js-action').contains('TriggerCallback')).toEqual(true);

    expect(target).toMatchSnapshot();
  });

  test('fail to open modal with disabled button', () => {
    const target = shallow(<ActionModal disabled={true} />);
    target.find('.js-open').simulate('click');
    expect(target.find('Modal')).toHaveProp('show', false);
    expect(target).toMatchSnapshot();
  });

  describe('with an onAction callback', () => {
    let target;
    beforeEach(() => {
      target = shallow(<ActionModal disabled={false} action="Next" />);
    });

    test('successful open modal with enabled button', () => {
      target.find('.js-open').simulate('click');
      expect(target.find('Modal')).toHaveProp('show', true);
      expect(target).toMatchSnapshot();
    });

    test('successful close modal', () => {
      target.find('.js-open').simulate('click');
      target.find('.js-close').simulate('click');
      expect(target.find('Modal')).toHaveProp('show', false);
      expect(target).toMatchSnapshot();
    });

    test('successful close modal on hide', () => {
      target.find('.js-open').simulate('click');
      target.find('Modal').simulate('hide');
      expect(target.find('Modal')).toHaveProp('show', false);
      expect(target).toMatchSnapshot();
    });
  });

  describe('with an onAction callback', () => {
    let target;
    let callback;
    beforeEach(() => {
      callback = jest.fn();
      target = shallow(<ActionModal disabled={false} onAction={callback} action="Next" />);
    });

    test('fail to close modal with not opened on action', () => {
      target.find('.js-action').simulate('click');
      expect(callback).not.toHaveBeenCalled();
      expect(target.find('Modal')).toHaveProp('show', false);
      expect(target).toMatchSnapshot();
    });

    test('successful close modal on action', () => {
      target.find('.js-open').simulate('click');
      target.find('.js-action').simulate('click');
      expect(callback).toHaveBeenCalled();
      expect(target.find('Modal')).toHaveProp('show', false);
      expect(target).toMatchSnapshot();
    });
  });
});
