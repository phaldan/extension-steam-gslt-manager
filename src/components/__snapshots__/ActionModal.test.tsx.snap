// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ActionModal> fail to open modal with disabled button 1`] = `
<span>
  <Button
    active={false}
    block={false}
    bsClass="btn"
    bsStyle="success"
    className="js-open"
    disabled={true}
    onClick={[Function]}
  >
     
  </Button>
  <Modal
    animation={true}
    autoFocus={true}
    backdrop={true}
    bsClass="modal"
    dialogComponentClass={[Function]}
    enforceFocus={true}
    keyboard={true}
    manager={
      ModalManager {
        "add": [Function],
        "containers": Array [],
        "data": Array [],
        "handleContainerOverflow": true,
        "hideSiblingNodes": true,
        "isTopModal": [Function],
        "modals": Array [],
        "remove": [Function],
      }
    }
    onHide={[Function]}
    renderBackdrop={[Function]}
    restoreFocus={true}
    show={false}
  >
    <ModalHeader
      bsClass="modal-header"
      closeButton={true}
      closeLabel="Close"
    >
      <ModalTitle
        bsClass="modal-title"
        componentClass="h4"
      />
    </ModalHeader>
    <ModalBody
      bsClass="modal-body"
      componentClass="div"
    />
  </Modal>
</span>
`;

exports[`<ActionModal> style and text customizations 1`] = `
<span>
  <Button
    active={false}
    block={false}
    bsClass="btn"
    bsSize="small"
    bsStyle="success"
    className="js-open"
    disabled={false}
    onClick={[Function]}
  >
    <Glyphicon
      bsClass="glyphicon"
      glyph="more"
    />
     
    Open
  </Button>
  <Modal
    animation={true}
    autoFocus={true}
    backdrop={true}
    bsClass="modal"
    dialogComponentClass={[Function]}
    enforceFocus={true}
    keyboard={true}
    manager={
      ModalManager {
        "add": [Function],
        "containers": Array [],
        "data": Array [],
        "handleContainerOverflow": true,
        "hideSiblingNodes": true,
        "isTopModal": [Function],
        "modals": Array [],
        "remove": [Function],
      }
    }
    onHide={[Function]}
    renderBackdrop={[Function]}
    restoreFocus={true}
    show={false}
  >
    <ModalHeader
      bsClass="modal-header"
      closeButton={true}
      closeLabel="Close"
    >
      <ModalTitle
        bsClass="modal-title"
        componentClass="h4"
      >
        TitleText
      </ModalTitle>
    </ModalHeader>
    <ModalBody
      bsClass="modal-body"
      componentClass="div"
    >
      <p>
        Lorem Ipsum
      </p>
    </ModalBody>
    <ModalFooter
      bsClass="modal-footer"
      componentClass="div"
    >
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="default"
        className="js-close"
        disabled={false}
        onClick={[Function]}
      >
        Close
      </Button>
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="primary"
        className="js-action"
        disabled={false}
        onClick={[Function]}
      >
        TriggerCallback
      </Button>
    </ModalFooter>
  </Modal>
</span>
`;

exports[`<ActionModal> with an onAction callback fail to close modal with not opened on action 1`] = `
<span>
  <Button
    active={false}
    block={false}
    bsClass="btn"
    bsStyle="success"
    className="js-open"
    disabled={false}
    onClick={[Function]}
  >
     
  </Button>
  <Modal
    animation={true}
    autoFocus={true}
    backdrop={true}
    bsClass="modal"
    dialogComponentClass={[Function]}
    enforceFocus={true}
    keyboard={true}
    manager={
      ModalManager {
        "add": [Function],
        "containers": Array [],
        "data": Array [],
        "handleContainerOverflow": true,
        "hideSiblingNodes": true,
        "isTopModal": [Function],
        "modals": Array [],
        "remove": [Function],
      }
    }
    onHide={[Function]}
    renderBackdrop={[Function]}
    restoreFocus={true}
    show={false}
  >
    <ModalHeader
      bsClass="modal-header"
      closeButton={true}
      closeLabel="Close"
    >
      <ModalTitle
        bsClass="modal-title"
        componentClass="h4"
      />
    </ModalHeader>
    <ModalBody
      bsClass="modal-body"
      componentClass="div"
    />
    <ModalFooter
      bsClass="modal-footer"
      componentClass="div"
    >
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="default"
        className="js-close"
        disabled={false}
        onClick={[Function]}
      >
        Abort
      </Button>
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="primary"
        className="js-action"
        disabled={false}
        onClick={[Function]}
      >
        Next
      </Button>
    </ModalFooter>
  </Modal>
</span>
`;

exports[`<ActionModal> with an onAction callback successful close modal 1`] = `
<span>
  <Button
    active={false}
    block={false}
    bsClass="btn"
    bsStyle="success"
    className="js-open"
    disabled={false}
    onClick={[Function]}
  >
     
  </Button>
  <Modal
    animation={true}
    autoFocus={true}
    backdrop={true}
    bsClass="modal"
    dialogComponentClass={[Function]}
    enforceFocus={true}
    keyboard={true}
    manager={
      ModalManager {
        "add": [Function],
        "containers": Array [],
        "data": Array [],
        "handleContainerOverflow": true,
        "hideSiblingNodes": true,
        "isTopModal": [Function],
        "modals": Array [],
        "remove": [Function],
      }
    }
    onHide={[Function]}
    renderBackdrop={[Function]}
    restoreFocus={true}
    show={false}
  >
    <ModalHeader
      bsClass="modal-header"
      closeButton={true}
      closeLabel="Close"
    >
      <ModalTitle
        bsClass="modal-title"
        componentClass="h4"
      />
    </ModalHeader>
    <ModalBody
      bsClass="modal-body"
      componentClass="div"
    />
    <ModalFooter
      bsClass="modal-footer"
      componentClass="div"
    >
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="default"
        className="js-close"
        disabled={false}
        onClick={[Function]}
      >
        Abort
      </Button>
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="primary"
        className="js-action"
        disabled={false}
        onClick={[Function]}
      >
        Next
      </Button>
    </ModalFooter>
  </Modal>
</span>
`;

exports[`<ActionModal> with an onAction callback successful close modal on action 1`] = `
<span>
  <Button
    active={false}
    block={false}
    bsClass="btn"
    bsStyle="success"
    className="js-open"
    disabled={false}
    onClick={[Function]}
  >
     
  </Button>
  <Modal
    animation={true}
    autoFocus={true}
    backdrop={true}
    bsClass="modal"
    dialogComponentClass={[Function]}
    enforceFocus={true}
    keyboard={true}
    manager={
      ModalManager {
        "add": [Function],
        "containers": Array [],
        "data": Array [],
        "handleContainerOverflow": true,
        "hideSiblingNodes": true,
        "isTopModal": [Function],
        "modals": Array [],
        "remove": [Function],
      }
    }
    onHide={[Function]}
    renderBackdrop={[Function]}
    restoreFocus={true}
    show={false}
  >
    <ModalHeader
      bsClass="modal-header"
      closeButton={true}
      closeLabel="Close"
    >
      <ModalTitle
        bsClass="modal-title"
        componentClass="h4"
      />
    </ModalHeader>
    <ModalBody
      bsClass="modal-body"
      componentClass="div"
    />
    <ModalFooter
      bsClass="modal-footer"
      componentClass="div"
    >
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="default"
        className="js-close"
        disabled={false}
        onClick={[Function]}
      >
        Abort
      </Button>
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="primary"
        className="js-action"
        disabled={false}
        onClick={[Function]}
      >
        Next
      </Button>
    </ModalFooter>
  </Modal>
</span>
`;

exports[`<ActionModal> with an onAction callback successful close modal on hide 1`] = `
<span>
  <Button
    active={false}
    block={false}
    bsClass="btn"
    bsStyle="success"
    className="js-open"
    disabled={false}
    onClick={[Function]}
  >
     
  </Button>
  <Modal
    animation={true}
    autoFocus={true}
    backdrop={true}
    bsClass="modal"
    dialogComponentClass={[Function]}
    enforceFocus={true}
    keyboard={true}
    manager={
      ModalManager {
        "add": [Function],
        "containers": Array [],
        "data": Array [],
        "handleContainerOverflow": true,
        "hideSiblingNodes": true,
        "isTopModal": [Function],
        "modals": Array [],
        "remove": [Function],
      }
    }
    onHide={[Function]}
    renderBackdrop={[Function]}
    restoreFocus={true}
    show={false}
  >
    <ModalHeader
      bsClass="modal-header"
      closeButton={true}
      closeLabel="Close"
    >
      <ModalTitle
        bsClass="modal-title"
        componentClass="h4"
      />
    </ModalHeader>
    <ModalBody
      bsClass="modal-body"
      componentClass="div"
    />
    <ModalFooter
      bsClass="modal-footer"
      componentClass="div"
    >
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="default"
        className="js-close"
        disabled={false}
        onClick={[Function]}
      >
        Abort
      </Button>
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="primary"
        className="js-action"
        disabled={false}
        onClick={[Function]}
      >
        Next
      </Button>
    </ModalFooter>
  </Modal>
</span>
`;

exports[`<ActionModal> with an onAction callback successful open modal with enabled button 1`] = `
<span>
  <Button
    active={false}
    block={false}
    bsClass="btn"
    bsStyle="success"
    className="js-open"
    disabled={false}
    onClick={[Function]}
  >
     
  </Button>
  <Modal
    animation={true}
    autoFocus={true}
    backdrop={true}
    bsClass="modal"
    dialogComponentClass={[Function]}
    enforceFocus={true}
    keyboard={true}
    manager={
      ModalManager {
        "add": [Function],
        "containers": Array [],
        "data": Array [],
        "handleContainerOverflow": true,
        "hideSiblingNodes": true,
        "isTopModal": [Function],
        "modals": Array [],
        "remove": [Function],
      }
    }
    onHide={[Function]}
    renderBackdrop={[Function]}
    restoreFocus={true}
    show={true}
  >
    <ModalHeader
      bsClass="modal-header"
      closeButton={true}
      closeLabel="Close"
    >
      <ModalTitle
        bsClass="modal-title"
        componentClass="h4"
      />
    </ModalHeader>
    <ModalBody
      bsClass="modal-body"
      componentClass="div"
    />
    <ModalFooter
      bsClass="modal-footer"
      componentClass="div"
    >
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="default"
        className="js-close"
        disabled={false}
        onClick={[Function]}
      >
        Abort
      </Button>
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="primary"
        className="js-action"
        disabled={false}
        onClick={[Function]}
      >
        Next
      </Button>
    </ModalFooter>
  </Modal>
</span>
`;
